plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts

```diff
@@ -0,0 +1,98 @@
+import { ExportProfilePluginMethod } from "@grouparoo/core";
+import { connect } from "../connect";
+import { addToList, removeFromList } from "./listMethods";
+
+export const exportProfile: ExportProfilePluginMethod = async ({
+  appGuid,
+  appOptions,
+  export: {
+    toDelete,
+    newProfileProperties,
+    oldProfileProperties,
+    newGroups,
+    oldGroups,
+  },
+}) => {
+  // if we received no mapped data... just exit
+  if (Object.keys(newProfileProperties).length === 0) {
+    return { success: true };
+  }
+  const client = await connect(appOptions);
+
+  const email = newProfileProperties["email"]; // this is how we will identify profiles
+  const currentEmail = oldProfileProperties["email"];
+  if (!email) {
+    throw new Error(`newProfileProperties[email] is a required mapping`);
+  }
+
+  // Remove email from profile properties to avoid duplicates.
+  delete oldProfileProperties.email;
+  delete newProfileProperties.email;
+
+  try {
+    if (toDelete) {
+      await client.users.delete(email);
+    } else {
+      // create the user and set properties
+      const deletePropertiesPayload = {};
+      const newPropertyKeys = Object.keys(newProfileProperties);
+      Object.keys(oldProfileProperties)
+        .filter((k) => !newPropertyKeys.includes(k))
+        .forEach((k) => (deletePropertiesPayload[k] = null));
+
+      const dataFields = Object.assign(
+        {},
+        newProfileProperties,
+        deletePropertiesPayload
+      );
+      const formattedDataFields = {};
+      for (const key of Object.keys(dataFields)) {
+        formattedDataFields[key] = formatVar(dataFields[key]);
+      }
+      const payload = Object.assign(
+        { email },
+        { dataFields: formattedDataFields }
+      );
+
+      await client.users.update(payload);
+
+      // add to lists
+      for (const groupToAdd of newGroups) {
+        await addToList(client, appGuid, appOptions, email, groupToAdd);
```

Comment:
does the response from `client.users.update(payload);` above have the listIds the user is in? You could use that here to only add if they aren't in the list (and remove if only if they are in the list).
This would save API calls.

Reply:
nope, the response shows only the status.

Reply:
It's interesting that the `getUser` (from tests) call seems to have that data.
There's some number (2+) of lists where making that call is the more efficient thing. But then there is also the timing issue of the user showing up. Not great.

Maybe it's actually better to put that back at the beginning and use that data when making the list. It could also be combined with the change email situation to see that they are already there.


```
  let existingUser = await client.users.get(email);
  if (currentEmail && email !== currentEmail) {
    const oldUser = await client.users.get(oldEmail);
    if (oldUser) {
      ....
      await client.users.updateEmail(emailPayload);
      existingUser = oldUser;
      existingUser.email = email;
    }
  }
```
This makes the switchover much smarter at the expense of these 2 calls. and one of them makes the update lists case possible much more efficent.

Then you'll have `existingUser` lower down to use if it's not `null`. use `existingUser.dataFields.emailListIds` to make add/remove smarter.